dora/core/client/fs/src/main/java/alluxio/client/file/dora/WorkerLocationPolicy.java

```diff
@@ -57,42 +63,109 @@ public WorkerLocationPolicy(int numVirtualNodes) {
+    }
 
+    /**
+     * Initializes or refreshes the worker list using the given list of workers and number of
+     * virtual nodes.
+     * <br>
+     * Thread safety:
+     * If called concurrently by two or more threads, only one of the callers will actually
+     * update the state of the hash provider using the worker list provided by that thread, and all
+     * others will not change the internal state of the hash provider.
+     */
     public void refresh(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
-      // check if we need to update worker info
-      if (mLastUpdatedTimestamp <= 0L
-          || System.currentTimeMillis() - mLastUpdatedTimestamp > WORKER_INFO_UPDATE_INTERVAL_MS) {
-        mNeedUpdate.set(true);
+      Preconditions.checkArgument(!workerInfos.isEmpty(),
+          "cannot refresh hash provider with empty worker list");
+      maybeInitialize(workerInfos, numVirtualNodes);
+      // check if the worker list has expired
+      long lastUpdateTs = mLastUpdatedTimestamp.get();
+      long currentTs = System.nanoTime();
+      if (currentTs - lastUpdateTs > mWorkerInfoUpdateIntervalNs) {
+        // use CAS to only allow one thread to actually update the timestamp
+        boolean casUpdated = mLastUpdatedTimestamp.compareAndSet(lastUpdateTs, currentTs);
+        // thread safety is valid provided that build() takes less than
+        // WORKER_INFO_UPDATE_INTERVAL_NS, so that before next update the current update has been
+        // finished
+        if (casUpdated) {
+          if (hasWorkerListChanged(workerInfos, mLastWorkerInfos)) {
+            mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
+            mLastWorkerInfos.clear();
+            mLastWorkerInfos.addAll(workerInfos);
+            mUpdateCount.incrementAndGet();
+          }
+        }
+        // else, do nothing and proceed with stale worker list. on next access, the worker list
+        // will have been updated by another thread
       }
-      // update worker info if needed
-      if (mNeedUpdate.compareAndSet(true, false)) {
-        if (isWorkerInfoUpdated(workerInfos, mLastWorkerInfos)) {
-          build(workerInfos, numVirtualNodes);
+    }
+
+    /**
+     * Lazily initializes the hash ring.
+     * Only one caller gets to initialize the map while all others are blocked.
+     * After the initialization, the map must not be null.
+     */
+    private void maybeInitialize(List<BlockWorkerInfo> workerInfos, int numVirtualNodes) {
+      if (mActiveNodesByConsistentHashing == null) {
+        synchronized (mInitLock) {
+          // only one thread should reach here
+          // test again to skip re-initialization
+          if (mActiveNodesByConsistentHashing == null) {
+            mActiveNodesByConsistentHashing = build(workerInfos, numVirtualNodes);
+            mLastWorkerInfos.clear();
+            mLastWorkerInfos.addAll(workerInfos);
```

Comment:
actually instead of clearing/adding, what about an AtomicRef to the new Set/List?

Reply:
I just feel like `final` with a mutable data structure is a bit inconsistent. And under Dora you might have 1000+ worker nodes, which makes a list/set generation non-trivial.